### PR TITLE
ARROW-7146: [R][CI] Various fixes and speedups for the R docker-compose setup

### PR DESCRIPTION
**Pull request by nealrichardson**
There's a lot of little things in here that have the effect of cutting 9 minutes off the build time and fixing a few bugs.

* Move R package dependency installation into the Dockerfile for better caching
* Sets MAKEFLAGS to the actual number of cores, not hard-coded to 8 (if this works out well I'll do the same to fix https://issues.apache.org/jira/browse/ARROW-6957)
* Removes hard-coded references to `bionic` so that R installation will work appropriately on other Ubuntu versions; also makes note of a dependency between OS version and availability of R versions for those who might try to run an older R version.
* Installs a lighter version of Romain's `decor` package, which cuts out dozens of unnecessary dependencies. I may remove this altogether, or at least flag it to be off on CI, because we don't need it there, it's just for developers.
* Skips vignette checks, which thereby avoid compiling and installing the R package an extra time (the check is pointless for us because there's no R code in the vignette)
* Removes an unnecessary `R CMD INSTALL`, which compiled the R bindings again (i.e., previously we compiled the R bindings 3 times, now we're down to 1)
* Turns off a check for valid URLs, saving about 30s
* Turns on more verbose error printing when there's a failure, rather than the default truncation at 13 lines.
* Uses the `rcmdcheck` package so we can fail the build on check Warnings (which cause CRAN failures). This was default on Travis.
* Adds some missing UTF-8 configuration to the dockerfile to fix some check Warnings
* Fixes some bad documentation crossreferences from #5859 that were causing other Warnings
* Correctly passes `-Werror` into the docker environment, which #5940 tried to add
* Restore the `TEST_R_WITH_ARROW` env var (lost in the move from Travis-CI) to make sure we're not just skipping all the tests

### CODE REVIEW

**Comment by nealrichardson** [comment]
Comparing with master, the Docker Run step is down from 20m to under 11m on this branch. I have the expected failure now, will now fix the bugs causing the Warnings (on master but not failing the build like they should).

**Comment by pitrou** [comment]
I'm excited by shorter build times @nealrichardson :-)

**Comment by nealrichardson** [comment]
Appveyor is green on my fork: https://ci.appveyor.com/project/nealrichardson/arrow/builds/29344837

File: ci/docker/linux-apt-r.dockerfile
Comment (kou) [inline ci/docker/linux-apt-r.dockerfile]: How about putting this into the above `apt-get install` instead of adding a new `apt-get install`?

File: ci/docker/linux-apt-r.dockerfile
Comment (kou) [inline ci/docker/linux-apt-r.dockerfile]: Could you sort this list alphabetically?

File: ci/scripts/r_deps.sh
Comment (kou) [inline ci/scripts/r_deps.sh]: Should we wait for https://github.com/romainfrancois/decor/pull/1 ?

Comment (nealrichardson) [reply ci/scripts/r_deps.sh]: I wouldn't let this hold up the PR. I'm happy to fold that change into a future patch, once Romain merges that.

An alternative would be to remove `decor` entirely because it's not used in CI; it's just installed here as convenience in case a (hypothetical) developer wanted to pull the docker image and edit the Rcpp code.

File: ci/docker/linux-apt-r.dockerfile
Comment (nealrichardson) [reply ci/docker/linux-apt-r.dockerfile]: I made it separate because the code comments implied that all that was copy-pasted from somewhere else, and I didn't want someone in the future to accidentally undo it, but sure, I can move it.

**Comment by nealrichardson** [comment]
Rust failures are network issues (or something), not related to this patch.

File: ci/scripts/r_deps.sh
Comment (kou) [reply ci/scripts/r_deps.sh]: OK.

**Review by kou — APPROVED** [review]
+1

**Review by fsaintjacques — COMMENTED** [review]
Tested locally and it worked,  thanks for doing this.

**Pull request closed without merging**
